rfd/0172-tbot-ssh-muxer.md

```diff
@@ -0,0 +1,183 @@
+state: draft
+---
+ 
+# RFD 172 - Long-lived Local `tbot` SSH Proxy
+
+## Required Approvers
+
+- Engineering: @rosstimothy || @espadolini
+
+## What
+
+Implement an optional local long-lived SSH connection proxy as part of `tbot`.
+This will be an optional feature that can be enabled via configuration and will
+not replace the existing SSH features of `tbot`.
+
+## Why
+
+Today, SSH proxying with OpenSSH and Machine ID relies on a short-lived `tbot`
+process being created for each SSH connection.
+
+This has a few drawbacks:
+
+- The `tbot` binary has a significant memory overhead, much of which is not
+  necessary for proxying the SSH connection. When connecting to a large number
+  of hosts, this overhead can become significant.
+- A process per connection, with each process having a number of tasks, places
+  the Linux scheduler under significant load.
+- Each `tbot` process must perform a number of initial startup tasks
+  individually, contributing to latency in establishing the SSH tunnel. Some of 
+  these tasks could be shared, or their results cached.
+- Monitoring a large number of short-lived `tbot` processes is difficult.
+- There is no easy way to implement global ratelimits and backoffs across all
+  the short-lived `tbot` processes. This makes gracefully handling upstream
+  outages problematic.
+
+## Details
+
+### Implementation
+
+We will introduce a long-lived service in `tbot`: `ssh-local-proxy`.
+
+Upon start, the service will write a SSH config to a configured destination.
+This SSH config should be used by OpenSSH to leverage the service.
+
+Once initialized, the service will listen on a Unix socket. Upon connection to this
+socket, the client will write a short message containing the target host and
+port. The service will then establish the connection, and then begin forwarding
+data between the socket and the upstream connection.
+
+The client will be configured as the `ProxyCommand`, and additionally
+`ProxyUseFdpass` will be enabled. The `ProxyUseFdPass` allows the `ProxyCommand`
+to return an FD via STDOUT using `sendmsg`, rather than the `ProxyCommand`
+actively forwarding the connection. This means the client can exit after the
+connection to the local proxy is established.
+
+This solution means that whilst a process will be spun up for each SSH
+connection, it will return after the connection is established. This reduces the
+long-term overhead associated with each connection and the impact on the Linux
+scheduler.
+
+#### Protocol
+
+A simple protocol will exist for interacting with the multiplexer:
+
+1. The server will open a Unix domain socket named `tbot_ssh_multiplexer.v1.sock`.
+2. Upon connection, the client will send a request indicating the target.
+  a. The request will be encoded in JSON and be followed by a NUL (ASCII/Unicode 0) character.
+  b. This JSON message will contain two fields:
+    i. `host`: The target host.
+    ii. `port`: The target host's port.
```

Comment:
What are we gaining with the JSON?

Reply:
I'm not extremely married to this decision - I went with JSON as it would make it trivial to add additional optional fields in the future as necessary without the need for breaking the API version. If there's some perceived risk with JSON, I'm more than happy to switch it back to just `host:port` terminated with NULL.